Change power numeric types from int to float for consistency

daniel-zullo-frequenz opened the pull request:
This modification reduces inconsistencies between different numeric types used to represent power values.

Fixes #28

daniel-zullo-frequenz committed:
Change power numeric types from int to float for consistency

This modification reduces inconsistencies between different
numeric types used to represent power values.

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz committed:
Use math.isclose() for floating-point numbers comparison

It is generally better to avoid using the '==' operator
for comparing floating-point numbers. Although it may not
be strictly necessary in some cases, using math.isclose()
is a safer approach.

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz committed:
Remove unnecessary ceil/floor conversions for power values

Since power types are now consistently defined as floats,
there is no need to convert power values to ints during
calculations.

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz commented on src/frequenz/sdk/actor/power_distributing/power_distributing.py:
```diff
@@ -351,7 +349,7 @@ async def _set_distributed_power(
         """
         tasks = {
-            inverter_id: asyncio.create_task(api.set_power(inverter_id, power))
+            inverter_id: asyncio.create_task(api.set_power(inverter_id, round(power)))
```
This is the only conversion from float to int currently required to set the power in the microgrid API (this will no longer needed once the microgrid API is updated to accept float type to set the power).
Is it fine to convert to int with round()? Or should it use something else like int(), floor() or ceil()?

sahas-subramanian-frequenz replied:
I think we should use `floor`,  because that's what the microgrid API will use when it is implemented.  For now, I think it might be better to do the conversion here:  https://github.com/frequenz-floss/frequenz-sdk-python/blob/c565ddf9e2e240c55e2873eb351daf4589e262db/src/frequenz/sdk/microgrid/client/_client.py#L589-L607

and call `floor` on both the sides.

daniel-zullo-frequenz replied:
You're right, the client seems to be the right place to do the conversion. And I've already updated to use floor() instead

daniel-zullo-frequenz commented on tests/power/test_distribution_algorithm.py:
```diff
@@ -72,6 +73,30 @@ def battery_msg(  # pylint: disable=too-many-arguments
     )
 
 
+def compare_dict(
```
This function is not needed as pytest.approx() can be used to compared dict with floating point values instead

mathias-baumann-frequenz reviewed (commented):
* In one commit you use `abs=1e-2` in another you use `abs=1e-9` why the difference?
* In the first commit you already replace a bunch of `==` comparisons with `isclose`, but in the second commit you explicitly say this is what the commit is doing. It would be more consistent to do it all in that commit then.
* Does it make sense to define our `1-e2` or `1-e9` (whichever we want to use) as a constant for easy reuse?

daniel-zullo-frequenz commented:
> In one commit you use abs=1e-2 in another you use abs=1e-9 why the difference?

`abs=1e-2` was only used in a test file, this was probably me being lazy to set all the digits required for the comparison but `1e-9` is generally used to compare floating point numbers for equality.

> In the first commit you already replace a bunch of == comparisons with isclose, but in the second commit you explicitly say this is what the commit is doing. It would be more consistent to do it all in that commit then.

yes, you're right. This should go in one commit


> Does it make sense to define our 1-e2 or 1-e9 (whichever we want to use) as a constant for easy reuse?

It does make sense. I am not sure if we can define a unique constant that can be accessed in the modules and tests. I'll check it out

leandro-lucarella-frequenz commented on src/frequenz/sdk/power/_distribution_algorithm.py:
```diff
@@ -396,24 +396,26 @@ def _greedy_distribute_remaining_power(
             Return the power for each inverter in given distribution.
         """
-        if remaining_power == 0:
+        if isclose(remaining_power, 0.0, abs_tol=1e-9):
```
I would add a comment saying why you are choosing that specific absolute tolerance (and not the default relative tolerance).

daniel-zullo-frequenz replied:
Added a brief comment and reference for the function you suggested in https://github.com/frequenz-floss/frequenz-sdk-python/pull/337#discussion_r1176616976

leandro-lucarella-frequenz commented on src/frequenz/sdk/power/_distribution_algorithm.py:
```diff
@@ -396,24 +396,26 @@ def _greedy_distribute_remaining_power(
 
         for inverter_id, power in distribution.items():
-            if remaining_power == 0 or power == 0:
+            if isclose(remaining_power, 0.0, abs_tol=1e-9) or isclose(
```
Maybe have a wrapper with these parameters so you have to write them and document them only once?

You can even use [`functools.partial`](https://docs.python.org/3/library/functools.html#functools.partial) instead of an explicit wrapper function.

```py
# We use this tolerance because X and Y
is_close_to_zero = partial(math.isclose, b=0.0, abs_tol=1e-9)

if is_close_to_zero(remaining_power) or is_close_to_zero(power):
    ...
```

daniel-zullo-frequenz replied:
This is great idea, I've implemented as suggested.

leandro-lucarella-frequenz reviewed (commented):
Besides the small comments, LGTM.

daniel-zullo-frequenz commented:
Updated based on suggestions from Mathias and Luca.

I've removed the tolerance in unit tests for comparing floating-point numbers for the function pytest.approx as in other tests in the SDK is used without tolerance.

daniel-zullo-frequenz reviewed (commented):
I need to add an entry to the release notes

leandro-lucarella-frequenz commented on src/frequenz/sdk/power/_distribution_algorithm.py:
```diff
@@ -5,13 +5,20 @@
+# and robustness for floating-point numbers comparisons close to zero. Note
+# that this is also the default value for the relative tolerance.
+# For more technical details, see https://peps.python.org/pep-0485/#behavior-near-zero
+_is_close_to_zero = partial(isclose, b=0.0, abs_tol=1e-9)
```
Super silly, and I don't think it is important to change now, but I would use the fully qualified name here (for both `math` and `functools`) for 4 reasons:

1. You only use these symbols once, so it is not even more typing
2. It is more clear when you read the code, as they are pretty generic names what you are importing
3. Avoid polluting the global namespace
4. Be more compliant with google style imports

It would be good to have this in mind for future PRs though.

daniel-zullo-frequenz replied:
Good points, I'll update it now as it is a simple change

leandro-lucarella-frequenz commented:
> I've removed the tolerance in unit tests for comparing floating-point numbers for the function pytest.approx as in other tests in the SDK is used without tolerance.

As long as tests are not flaky, LGTM.

leandro-lucarella-frequenz commented:
> I need to add an entry to the release notes

I'll approve after this is done.

daniel-zullo-frequenz commented:
Added an entry to the release notes and updated imports to use the fully qualified name both math and functools.

daniel-zullo-frequenz commented:
Rebased onto latest v0.x.x

leandro-lucarella-frequenz commented on RELEASE_NOTES.md:
```diff
@@ -6,7 +6,9 @@
-<!-- Here goes notes on how to upgrade from previous versions, including deprecations and what they should be replaced with --> 
+* Improved consistency in `frequenz.sdk.actor.power_distributing` package by unifying the representation of power values across different numeric types:
+  * The `power` attribute of the `Request` class has been updated from an integer to a float to better represent values of power. This change also affects internal variables and checks related to power values, which have been updated accordingly.
+  * The `set_power()` method of both the `MicrogridApiClient` and `MicrogridGrpcClient` classes in import `frequenz.sdk.microgrid` package now expect a float value for the `power_w` parameter instead of an integer, to align with the new power value representation in the power distributing package.
```
```suggestion
* Now `float` is used everywhere for representing power (before power metrics were `float` but setting power was done using `int`).
  - `frequenz.sdk.actor.power_distributing`: the `power` attribute of the `Request` class has been updated from `int` to a `float`.
  * `frequenz.sdk.microgrid`: the `set_power()` method of both the `MicrogridApiClient` and `MicrogridGrpcClient` classes now expect a `float` value for the `power_w` parameter instead of `int`.
```

A few recommendations on release notes entries that I think it makes it more readable and useful:

* Summarize the main change and how users should upgrade.
* Make the affected package and classes very clear and precisely what was changed and how.
* Don't talk about internals, this notes are targeted to users, they don't care if the internal representation was changed or not, nor how. Those observations should go in commit messages.
* Try to put yourself in the shoes of the user that needs to update the SDK dependency and upgrade their project, what is useful information and what isn't?

daniel-zullo-frequenz replied:
Makes sense, thanks for the suggestions. Updated as suggested.

leandro-lucarella-frequenz commented on examples/power_distribution.py:
```diff
@@ -82,10 +82,10 @@ async def run(self) -> None:
             _logger.debug("Avg power %d", avg_power)
             if avg_power > 30000:
                 # Charge
-                power_to_set: int = 10000
+                power_to_set: float = 10000.0
```
Even when not related to this PR, I recommend *declaring* the type outside of the `if` here to make it more clear:

```py
            power_to_set: float
            if avg_power > 30000:
                # Charge
                power_to_set = 10000.0
```

daniel-zullo-frequenz replied:
Updated

leandro-lucarella-frequenz reviewed (commented):
Besides the small comments, LGTM. There was also the suggestion to move `_is_close_to_zero()` to `_internal` that @sahas-subramanian-frequenz mentioned.

daniel-zullo-frequenz committed:
Rename file names for internal modules

To make the file names in internal consistent.

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz commented:
Updated based on suggestions from Luca and Sahas.

Added a new commit to rename files in `internal` so that all the files there are consistently named with the prefix `_`

leandro-lucarella-frequenz commented on src/frequenz/sdk/_internal/_math.py:
```diff
@@ -0,0 +1,13 @@
+# and robustness for floating-point numbers comparisons close to zero. Note
+# that this is also the default value for the relative tolerance.
+# For more technical details, see https://peps.python.org/pep-0485/#behavior-near-zero
+_is_close_to_zero = functools.partial(math.isclose, b=0.0, abs_tol=1e-9)
```
If this is now in a separate module I would make it regular function that calls `math.isclose()`, so it can be properly documented (probably including what's in the comments as part of the documentation). Also, remove the leading `_`, as this is a **public** function (in a **private** module).

daniel-zullo-frequenz replied:
True, the functools was helpful to use it in the previous case but in this independent private module makes sense to have a regular function. I have already updated it accordingly

leandro-lucarella-frequenz reviewed (commented):
Just one last small thing from my side :grimacing:

daniel-zullo-frequenz commented:
Updated based on latest suggestion from Luca

sahas-subramanian-frequenz reviewed (approved):
:tada:

daniel-zullo-frequenz commented:
Thanks Sahas! I'll wait for Luca's approval to merge it as he's requested a change (already addressed)

leandro-lucarella-frequenz reviewed (approved):
:shipit:
